src/integrationTest/java/org/opensearch/security/api/AbstractApiIntegrationTest.java

```diff
@@ -0,0 +1,300 @@
+ *
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+package org.opensearch.security.api;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.StringJoiner;
+
+import com.carrotsearch.randomizedtesting.RandomizedTest;
+import com.carrotsearch.randomizedtesting.annotations.ThreadLeakScope;
+import com.google.common.collect.ImmutableMap;
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import org.apache.commons.io.FileUtils;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.awaitility.Awaitility;
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.runner.RunWith;
+
+import org.opensearch.common.CheckedConsumer;
+import org.opensearch.common.xcontent.XContentFactory;
+import org.opensearch.core.xcontent.ToXContent;
+import org.opensearch.security.ConfigurationFiles;
+import org.opensearch.security.securityconf.impl.CType;
+import org.opensearch.test.framework.TestSecurityConfig;
+import org.opensearch.test.framework.certificate.CertificateData;
+import org.opensearch.test.framework.cluster.ClusterManager;
+import org.opensearch.test.framework.cluster.LocalCluster;
+import org.opensearch.test.framework.cluster.TestRestClient;
+
+import static org.hamcrest.Matchers.equalTo;
+import static org.opensearch.security.CrossClusterSearchTests.PLUGINS_SECURITY_RESTAPI_ROLES_ENABLED;
+import static org.opensearch.security.DefaultObjectMapper.objectMapper;
+import static org.opensearch.security.support.ConfigConstants.SECURITY_ALLOW_DEFAULT_INIT_SECURITYINDEX;
+import static org.opensearch.security.support.ConfigConstants.SECURITY_RESTAPI_ADMIN_ENABLED;
+
+@ThreadLeakScope(ThreadLeakScope.Scope.NONE)
+@RunWith(com.carrotsearch.randomizedtesting.RandomizedRunner.class)
+public abstract class AbstractApiIntegrationTest extends RandomizedTest {
+
+    private static final Logger LOGGER = LogManager.getLogger(TestSecurityConfig.class);
+
+    public static final String NEW_USER = "new-user";
+
+    public static final String REST_ADMIN_USER = "rest-api-admin";
+
+    public static final String ADMIN_USER_NAME = "admin";
+
+    public static final String DEFAULT_PASSWORD = "secret";
+
+    public static final String EMPTY_BODY = "{}";
+
+    public final static Path CONFIGURATION_FOLDER = ConfigurationFiles.createConfigurationDirectory();
+
+    public static Map<String, Object> clusterSettings;
+
+    public static Set<TestSecurityConfig.User> users;
+
+    public static Set<TestSecurityConfig.Role> roles;
+
+    public static Set<TestSecurityConfig.RoleMapping> rolesMappings;
+
+    public static Set<TestSecurityConfig.ActionGroup> actionGroups;
+
+    public static LocalCluster localCluster;
+
+    static class Patch {
```

Comment:
In the new ConfigUpgradeApiAction and the AbstractApiAction patch operations are reference, it might be worthwhile to push these defintions into the product code. 

https://github.com/opensearch-project/security/blob/a731e62f8a52420645473c8279081ae46bfb86b2/src/main/java/org/opensearch/security/dlic/rest/api/ConfigUpgradeApiAction.java#L236-L262

Reply:
So far not sure do we need it or not :-)

Reply:
I'm fine either way.